tests/kernel/mem_protect/mem_protect/src/mem_domain.c

```diff
@@ -594,3 +625,229 @@ void test_mem_domain_destroy(void)
 	}
 }
+
+/**
+ * @brief Test access memory domain APIs allowed to supervisor threads only
+ *
+ * @details
+ * - Create a test case to use a memory domain APIs k_mem_domain_init()
+ *   and k_mem_domain_add_partition()
+ * - Run that test in kernel mode -no errors should happen.
+ * - Run that test in user mode -Zephyr fatal error will happen (reason 0),
+ *   because user threads can't use memory domain APIs.
+ * - At the same time test that system support the definition of memory domains.
+ *
+ * @ingroup kernel_memprotect_tests
+ *
+ * @see k_mem_domain_init(), k_mem_domain_add_partition()
+ */
+void test_mem_domain_api_kernel_thread_only(void)
+{
+	valid_fault = true;
+
+	k_mem_domain_init(&domain0, 0, NULL);
+	k_mem_domain_add_partition(&domain0, &part0);
+}
+
+void user_handler_func(void *p1, void *p2, void *p3)
+{
+	/* Read the partition */
+	u8_t read_data = buf1[0];
+
+	/* Just to avoid compiler warning */
+	(void) read_data;
+
+	/* Writing to the partition, this should generate fault
+	 * as the partition has read only permission for
+	 * user threads
+	 */
+	buf1[0] = 10U;
+	k_sem_give(&sync_sem);
+}
+
+/**
+ * @brief Test system auto determine memory partition base addresses and sizes
+ *
+ * @details
+ * - Ensure that system automatically determine memory partition
+ *   base address and size according to its content correctly by taking
+ *   memory partition's size and compare it with size of content.
+ * - If size of memory partition is equal to the size of content -test passes.
+ * - If possible to take base address of the memory partition -test passes.
+ * - Test that system supports the definition of memory partitions and it has
+ *   a starting address and a size.
+ * - Test that OS support removing thread from it is domain assignment.
+ *
+ * @ingroup kernel_memprotect_tests
+ */
+void test_mem_part_auto_determ_size(void)
+{
+	k_sem_init(&sync_sem, SYNC_SEM_INIT_COUNT, SYNC_SEM_MAX_COUNT);
+	k_thread_access_grant(&user_thread0, &sync_sem);
+	zassert_true(part1.size == MEM_REGION_ALLOC, "Size of memory partition"
+			" determined not correct according to its content");
+	zassert_true(part1.start, "Base address of memory partition"
+			" not determined at build time");
+
+	k_mem_domain_init(&domain1, ARRAY_SIZE(app1_parts), app1_parts);
+	k_mem_domain_remove_thread(k_current_get());
+	usr_tid0 = k_thread_create(&user_thread0, user_thread0_stack,
+					K_THREAD_STACK_SIZEOF(
+						user_thread0_stack),
+					user_handler_func,
+					NULL, NULL, NULL,
+					PRIORITY,
+					K_USER, K_NO_WAIT);
+	k_mem_domain_add_thread(&domain1, usr_tid0);
+	k_sem_take(&sync_sem, K_FOREVER);
+}
+
+/**
+ * @brief Test partitions sized per the constraints of the MMU hardware
+ *
+ * @details
+ * - Test that system automatically determine memory partition size according
+ *   to the constraints of the platform's MMU hardware.
+ * - Different platforms like x86, ARM and ARC have different MMU hardware for
```

Comment:
No, ARC and some ARM do not have MMU at all, they have MPU instead.

Reply:
I will change comment and replace with "MPU".